lib/Dialect/FIRRTL/Transforms/InferWidths.cpp

```diff
@@ -0,0 +1,752 @@
+/// merely there for show and ease of use.
+struct BinaryExpr : public Expr {
+  bool operator==(const BinaryExpr &other) const {
+    return kind == other.kind && lhs() == other.lhs() && rhs() == other.rhs();
+  }
+  llvm::hash_code hash_value() const {
+    return llvm::hash_combine(Expr::hash_value(), *args);
+  }
+  Expr *lhs() const { return args[0]; }
+  Expr *rhs() const { return args[1]; }
+  iterator begin() const { return args; }
+  iterator end() const { return args + 2; }
+
+  /// The child expressions.
+  Expr *const args[2];
+
+protected:
+  BinaryExpr(Kind kind, Expr *lhs, Expr *rhs) : Expr(kind), args{lhs, rhs} {}
+};
+
+/// Helper class to CRTP-derive common functions.
+template <class DerivedT, Expr::Kind DerivedKind>
+struct BinaryExprBase : public BinaryExpr {
+  template <typename... Args>
+  BinaryExprBase(Args &&...args)
+      : BinaryExpr(DerivedKind, std::forward<Args>(args)...) {}
+  static bool classof(const Expr *e) { return e->kind == DerivedKind; }
+};
+
+/// An addition.
+struct AddExpr : public BinaryExprBase<AddExpr, Expr::KindAdd> {
+  using BinaryExprBase::BinaryExprBase;
+  void print(llvm::raw_ostream &os) const {
+    os << "(" << lhs() << " + " << rhs() << ")";
+  }
+};
+
+/// The maximum of two expressions.
+struct MaxExpr : public BinaryExprBase<MaxExpr, Expr::KindMax> {
+  using BinaryExprBase::BinaryExprBase;
+  void print(llvm::raw_ostream &os) const {
+    os << "max(" << lhs() << ", " << rhs() << ")";
+  }
+};
+
+/// The minimum of two expressions.
+struct MinExpr : public BinaryExprBase<MinExpr, Expr::KindMin> {
+  using BinaryExprBase::BinaryExprBase;
+  void print(llvm::raw_ostream &os) const {
+    os << "min(" << lhs() << ", " << rhs() << ")";
+  }
+};
+
+void Expr::print(llvm::raw_ostream &os) const {
+  TypeSwitch<const Expr *>(this)
+      .Case<RootExpr, VarExpr, KnownExpr, AddExpr, MaxExpr, MinExpr>(
+          [&](auto *e) { e->print(os); });
+}
+
+Expr::iterator Expr::begin() const {
+  return TypeSwitch<const Expr *, Expr::iterator>(this)
+      .Case<RootExpr, VarExpr, KnownExpr, AddExpr, MaxExpr, MinExpr>(
+          [&](auto *e) { return e->begin(); });
+}
+
+Expr::iterator Expr::end() const {
+  return TypeSwitch<const Expr *, Expr::iterator>(this)
+      .Case<RootExpr, VarExpr, KnownExpr, AddExpr, MaxExpr, MinExpr>(
+          [&](auto *e) { return e->end(); });
+}
+
+} // namespace
+
+//===----------------------------------------------------------------------===//
+// GraphTraits on constraint expressions
+//===----------------------------------------------------------------------===//
+
+namespace llvm {
+template <>
+struct GraphTraits<Expr *> {
+  using ChildIteratorType = Expr::iterator;
+  using NodeRef = Expr *;
+
+  static NodeRef getEntryNode(NodeRef N) { return N; }
+  static inline ChildIteratorType child_begin(NodeRef N) { return N->begin(); }
+  static inline ChildIteratorType child_end(NodeRef N) { return N->end(); }
```

Comment:
I would use a lower-case name like `node`.  This type of name is common in LLVM but I think its a convention that MLIR diverges from.

Reply:
Right.  LLVM's capitalization is a mess, MLIR is much more consistent.  We should align with MLIR.  More than you ever wanted to know:
https://mlir.llvm.org/getting_started/DeveloperGuide/
https://llvm.org/docs/Proposals/VariableNames.html